File: swift-bridge/Sources/CryptoTokenKitBridge/SmartCard.swift
import CryptoTokenKit
import Foundation

@_cdecl("ctk_smart_card_slot_name")
public func ctk_smart_card_slot_name(
    _ cardPtr: UnsafeMutableRawPointer?,
    _ errorOut: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UnsafeMutablePointer<CChar>? {
    guard let card = ctkBorrow(cardPtr, as: TKSmartCard.self, errorOut) else { return nil }
    return ctkCString(card.slot.name)
}

@_cdecl("ctk_smart_card_valid")
public func ctk_smart_card_valid(
    _ cardPtr: UnsafeMutableRawPointer?,
    _ errorOut: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Bool {
    guard let card = ctkBorrow(cardPtr, as: TKSmartCard.self, errorOut) else { return false }
    return card.isValid
}

@_cdecl("ctk_smart_card_allowed_protocols")
public func ctk_smart_card_allowed_protocols(
    _ cardPtr: UnsafeMutableRawPointer?,
    _ errorOut: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UInt32 {
    guard let card = ctkBorrow(cardPtr, as: TKSmartCard.self, errorOut) else { return 0 }
    return UInt32(card.allowedProtocols.rawValue)
}

@_cdecl("ctk_smart_card_set_allowed_protocols")
public func ctk_smart_card_set_allowed_protocols(
    _ cardPtr: UnsafeMutableRawPointer?,
    _ protocols: UInt32,
    _ errorOut: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) {
    guard let card = ctkBorrow(cardPtr, as: TKSmartCard.self, errorOut) else { return }
    card.allowedProtocols = TKSmartCardProtocol(rawValue: UInt(protocols))
}

@_cdecl("ctk_smart_card_current_protocol")
public func ctk_smart_card_current_protocol(
    _ cardPtr: UnsafeMutableRawPointer?,
    _ errorOut: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UInt32 {
    guard let card = ctkBorrow(cardPtr, as: TKSmartCard.self, errorOut) else { return 0 }
    return UInt32(card.currentProtocol.rawValue)
}

@_cdecl("ctk_smart_card_sensitive")
public func ctk_smart_card_sensitive(
    _ cardPtr: UnsafeMutableRawPointer?,
    _ errorOut: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Bool {
    guard let card = ctkBorrow(cardPtr, as: TKSmartCard.self, errorOut) else { return false }
    return card.isSensitive
}

@_cdecl("ctk_smart_card_set_sensitive")
public func ctk_smart_card_set_sensitive(
    _ cardPtr: UnsafeMutableRawPointer?,
    _ sensitive: Bool,
    _ errorOut: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) {
    guard let card = ctkBorrow(cardPtr, as: TKSmartCard.self, errorOut) else { return }
    card.isSensitive = sensitive
}

@_cdecl("ctk_smart_card_cla")
public func ctk_smart_card_cla(
    _ cardPtr: UnsafeMutableRawPointer?,
    _ errorOut: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UInt8 {
    guard let card = ctkBorrow(cardPtr, as: TKSmartCard.self, errorOut) else { return 0 }
    return card.cla
}

@_cdecl("ctk_smart_card_set_cla")
public func ctk_smart_card_set_cla(
    _ cardPtr: UnsafeMutableRawPointer?,
    _ cla: UInt8,
    _ errorOut: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) {
    guard let card = ctkBorrow(cardPtr, as: TKSmartCard.self, errorOut) else { return }
    card.cla = cla
}

@_cdecl("ctk_smart_card_use_extended_length")
public func ctk_smart_card_use_extended_length(
    _ cardPtr: UnsafeMutableRawPointer?,
    _ errorOut: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Bool {
    guard let card = ctkBorrow(cardPtr, as: TKSmartCard.self, errorOut) else { return false }
    return card.useExtendedLength
}

@_cdecl("ctk_smart_card_set_use_extended_length")
public func ctk_smart_card_set_use_extended_length(
    _ cardPtr: UnsafeMutableRawPointer?,
    _ enabled: Bool,
    _ errorOut: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) {
    guard let card = ctkBorrow(cardPtr, as: TKSmartCard.self, errorOut) else { return }
    card.useExtendedLength = enabled
}

@_cdecl("ctk_smart_card_use_command_chaining")
public func ctk_smart_card_use_command_chaining(
    _ cardPtr: UnsafeMutableRawPointer?,
    _ errorOut: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Bool {
    guard let card = ctkBorrow(cardPtr, as: TKSmartCard.self, errorOut) else { return false }
    return card.useCommandChaining
}

@_cdecl("ctk_smart_card_set_use_command_chaining")
public func ctk_smart_card_set_use_command_chaining(
    _ cardPtr: UnsafeMutableRawPointer?,
    _ enabled: Bool,
    _ errorOut: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) {
    guard let card = ctkBorrow(cardPtr, as: TKSmartCard.self, errorOut) else { return }
    card.useCommandChaining = enabled
}

@_cdecl("ctk_smart_card_context_json")
public func ctk_smart_card_context_json(
    _ cardPtr: UnsafeMutableRawPointer?,
    _ errorOut: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UnsafeMutablePointer<CChar>? {
    guard let card = ctkBorrow(cardPtr, as: TKSmartCard.self, errorOut) else { return nil }
    if let context = card.context as? CTKRustContextBox {
        return ctkCString(context.json)
    }
    if let context = card.context as? String {
        return ctkCString(context)
    }
    return nil
}

@_cdecl("ctk_smart_card_set_context_json")
public func ctk_smart_card_set_context_json(
    _ cardPtr: UnsafeMutableRawPointer?,
    _ json: UnsafePointer<CChar>?,
    _ hasJSON: Bool,
    _ errorOut: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    guard let cardPtr else {
        ctkWriteError(errorOut, "missing smart-card handle")
        return CTK_INVALID_ARGUMENT
    }
    guard let card = ctkBorrow(cardPtr, as: TKSmartCard.self, errorOut) else { return CTK_INVALID_ARGUMENT }
    if hasJSON {
        guard let json else {
            ctkWriteError(errorOut, "missing smart-card context JSON")
            return CTK_INVALID_ARGUMENT
        }
        card.context = CTKRustContextBox(json: String(cString: json))
    } else {
        card.context = nil
    }
    return CTK_OK
}

@_cdecl("ctk_smart_card_begin_session")
public func ctk_smart_card_begin_session(
    _ cardPtr: UnsafeMutableRawPointer?,
    _ errorOut: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    guard let cardPtr else {
        ctkWriteError(errorOut, "missing smart-card handle")
        return CTK_INVALID_ARGUMENT
    }

    guard let card = ctkBorrow(cardPtr, as: TKSmartCard.self, errorOut) else { return CTK_INVALID_ARGUMENT }
    let pending = CTKPendingReply<(Bool, Error?)>()
    card.beginSession { success, error in
        if !pending.complete((success, error)), success {
            card.endSession()
        }
    }
    guard let (success, error) = pending.wait(seconds: 30) else {
        ctkWriteError(errorOut, "timed out waiting for smart-card session")
        return CTK_TIMED_OUT
    }
    guard success else {
        ctkWriteNSError(errorOut, fallback: "failed to begin smart-card session", error: error)
        return error.map(ctkStatus(from:)) ?? CTK_FRAMEWORK_ERROR
    }
    return CTK_OK
}

@_cdecl("ctk_smart_card_transmit_request_json")
public func ctk_smart_card_transmit_request_json(
    _ cardPtr: UnsafeMutableRawPointer?,
    _ requestPtr: UnsafePointer<UInt8>?,
    _ requestLen: Int,
    _ outReplyJSON: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>,
    _ errorOut: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    outReplyJSON.pointee = nil
    guard let cardPtr else {
        ctkWriteError(errorOut, "missing smart-card handle")
        return CTK_INVALID_ARGUMENT
    }
    guard let requestPtr else {
        ctkWriteError(errorOut, "missing smart-card request")
        return CTK_INVALID_ARGUMENT
    }

    guard let card = ctkBorrow(cardPtr, as: TKSmartCard.self, errorOut) else { return CTK_INVALID_ARGUMENT }
    let request = Data(bytes: requestPtr, count: requestLen)
    let pending = CTKPendingReply<(Data?, Error?)>()
    card.transmit(request) { response, error in
        _ = pending.complete((response, error))
    }
    guard let (replyData, replyError) = pending.wait(seconds: 30) else {
        ctkWriteError(errorOut, "timed out waiting for smart-card transmit")
        return CTK_TIMED_OUT
    }
    if let replyError {
        ctkWriteNSError(errorOut, fallback: "smart-card transmit failed", error: replyError)
        return ctkStatus(from: replyError)
    }
    outReplyJSON.pointee = ctkCString(ctkJSONString([UInt8](replyData ?? Data())))
    return CTK_OK
}

@_cdecl("ctk_smart_card_end_session")
public func ctk_smart_card_end_session(
    _ cardPtr: UnsafeMutableRawPointer?,
    _ errorOut: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) {
    guard let card = ctkBorrow(cardPtr, as: TKSmartCard.self, errorOut) else { return }
    card.endSession()
}

@_cdecl("ctk_smart_card_send_ins")
public func ctk_smart_card_send_ins(
    _ cardPtr: UnsafeMutableRawPointer?,
    _ ins: UInt8,
    _ p1: UInt8,
    _ p2: UInt8,
    _ dataPtr: UnsafePointer<UInt8>?,
    _ dataLen: Int,
    _ hasLE: Bool,
    _ le: Int,
    _ outReplyJSON: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>,
    _ errorOut: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    outReplyJSON.pointee = nil
    guard let cardPtr else {
        ctkWriteError(errorOut, "missing smart-card handle")
        return CTK_INVALID_ARGUMENT
    }

    guard !hasLE || (0...65536).contains(le) else {
        ctkWriteError(errorOut, "le must be between 0 and 65536")
        return CTK_INVALID_ARGUMENT
    }

    guard let card = ctkBorrow(cardPtr, as: TKSmartCard.self, errorOut) else { return CTK_INVALID_ARGUMENT }
    let requestData = dataPtr.map { Data(bytes: $0, count: dataLen) }
    let expectedLength: Int? = hasLE ? le : nil

    let sessionStatus = ctk_smart_card_begin_session(cardPtr, errorOut)
    if sessionStatus != CTK_OK {
        return sessionStatus
    }
    defer { card.endSession() }

    do {
        let reply = try card.send(ins: ins, p1: p1, p2: p2, data: requestData, le: expectedLength)
        outReplyJSON.pointee = ctkCString(ctkJSONString([
            "data": [UInt8](reply.response),
            "statusWord": reply.sw,
        ]))
        return CTK_OK
    } catch {
        ctkWriteNSError(errorOut, fallback: "CryptoTokenKit APDU exchange failed", error: error)
        return ctkStatus(from: error)
    }
}

File: src/smart_card.rs
use core::ffi::c_void;
use std::ptr;

use serde::{Deserialize, Serialize};

use crate::error::CryptoTokenKitError;
use crate::ffi;
use crate::private::{decode_json, status_result};
use crate::smart_card_atr::SmartCardProtocol;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
/// Result payload returned from `TKSmartCard.send(ins:p1:p2:data:le:)`.
pub struct ApduResponse {
    #[serde(default, with = "serde_bytes")]
    /// Serialized field bridged from `TKSmartCard.send(ins:p1:p2:data:le:)`.
    pub data: Vec<u8>,
    /// Serialized field bridged from `TKSmartCard.send(ins:p1:p2:data:le:)`.
    pub status_word: u16,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[repr(i32)]
/// Mirrors the character-set choices stored by `TKSmartCardPINFormat`.
pub enum SmartCardPinCharset {
    /// Variant bridged from `TKSmartCardPINFormat`.
    Numeric = 0,
    /// Variant bridged from `TKSmartCardPINFormat`.
    Alphanumeric = 1,
    /// Variant bridged from `TKSmartCardPINFormat`.
    UpperAlphanumeric = 2,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[repr(i32)]
/// Mirrors the encoding choices stored by `TKSmartCardPINFormat`.
pub enum SmartCardPinEncoding {
    /// Variant bridged from `TKSmartCardPINFormat`.
    Binary = 0,
    /// Variant bridged from `TKSmartCardPINFormat`.
    Ascii = 1,
    /// Variant bridged from `TKSmartCardPINFormat`.
    Bcd = 2,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[repr(i32)]
/// Mirrors the justification options stored by `TKSmartCardPINFormat`.
pub enum SmartCardPinJustification {
    /// Variant bridged from `TKSmartCardPINFormat`.
    Left = 0,
    /// Variant bridged from `TKSmartCardPINFormat`.
    Right = 1,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
/// Wraps the completion flags used by `CryptoTokenKit` PIN interactions.
pub struct SmartCardPinCompletion(
    /// Serialized field bridged from `CryptoTokenKit PIN interaction flags`.
    pub u32,
);

impl SmartCardPinCompletion {
    /// Wraps the corresponding `CryptoTokenKit PIN interaction flags` operation.
    pub const MAX_LENGTH: Self = Self(1 << 0);
    /// Wraps the corresponding `CryptoTokenKit PIN interaction flags` operation.
    pub const KEY: Self = Self(1 << 1);
    /// Wraps the corresponding `CryptoTokenKit PIN interaction flags` operation.
    pub const TIMEOUT: Self = Self(1 << 2);

    #[must_use]
    /// Wraps the corresponding `CryptoTokenKit PIN interaction flags` operation.
    pub const fn bits(self) -> u32 {
        self.0
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
/// Wraps the confirmation flags used by `CryptoTokenKit` PIN-change interactions.
pub struct SmartCardPinConfirmation(
    /// Serialized field bridged from `CryptoTokenKit PIN change interaction flags`.
    pub u32,
);

impl SmartCardPinConfirmation {
    /// Wraps the corresponding `CryptoTokenKit PIN change interaction flags` operation.
    pub const NONE: Self = Self(0);
    /// Wraps the corresponding `CryptoTokenKit PIN change interaction flags` operation.
    pub const NEW: Self = Self(1 << 0);
    /// Wraps the corresponding `CryptoTokenKit PIN change interaction flags` operation.
    pub const CURRENT: Self = Self(1 << 1);

    #[must_use]
    /// Wraps the corresponding `CryptoTokenKit PIN change interaction flags` operation.
    pub const fn bits(self) -> u32 {
        self.0
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
/// Snapshot of the `TKSmartCardPINFormat` configuration.
pub struct SmartCardPinFormat {
    /// Serialized field bridged from `TKSmartCardPINFormat`.
    pub charset: SmartCardPinCharset,
    /// Serialized field bridged from `TKSmartCardPINFormat`.
    pub encoding: SmartCardPinEncoding,
    /// Serialized field bridged from `TKSmartCardPINFormat`.
    pub min_pin_length: i64,
    /// Serialized field bridged from `TKSmartCardPINFormat`.
    pub max_pin_length: i64,
    /// Serialized field bridged from `TKSmartCardPINFormat`.
    pub pin_block_byte_length: i64,
    /// Serialized field bridged from `TKSmartCardPINFormat`.
    pub pin_justification: SmartCardPinJustification,
    /// Serialized field bridged from `TKSmartCardPINFormat`.
    pub pin_bit_offset: i64,
    /// Serialized field bridged from `TKSmartCardPINFormat`.
    pub pin_length_bit_offset: i64,
    /// Serialized field bridged from `TKSmartCardPINFormat`.
    pub pin_length_bit_size: i64,
}

impl Default for SmartCardPinFormat {
    fn default() -> Self {
        Self {
            charset: SmartCardPinCharset::Numeric,
            encoding: SmartCardPinEncoding::Ascii,
            min_pin_length: 4,
            max_pin_length: 8,
            pin_block_byte_length: 8,
            pin_justification: SmartCardPinJustification::Left,
            pin_bit_offset: 0,
            pin_length_bit_offset: 0,
            pin_length_bit_size: 0,
        }
    }
}

/// Wraps `TKSmartCard`.
pub struct SmartCard {
    raw: *mut c_void,
}

impl SmartCard {
    #[must_use]
    pub(crate) const fn from_raw(raw: *mut c_void) -> Self {
        Self { raw }
    }

    #[must_use]
    pub(crate) const fn raw(&self) -> *mut c_void {
        self.raw
    }

    /// Wraps the corresponding `TKSmartCard` operation.
    pub fn slot_name(&self) -> Result<String, CryptoTokenKitError> {
        let ptr = unsafe { ffi::smart_card::ctk_smart_card_slot_name(self.raw) };
        if ptr.is_null() {
            return Err(CryptoTokenKitError::FrameworkError(
                "Swift bridge returned a null slot name".into(),
            ));
        }
        Ok(crate::error::take_owned_c_string(ptr))
    }

    #[must_use]
    /// Wraps the corresponding `TKSmartCard` operation.
    pub fn valid(&self) -> bool {
        unsafe { ffi::smart_card::ctk_smart_card_valid(self.raw) }
    }

    #[must_use]
    /// Wraps the corresponding `TKSmartCard` operation.
    pub fn allowed_protocols(&self) -> SmartCardProtocol {
        SmartCardProtocol::from_bits(unsafe {
            ffi::smart_card::ctk_smart_card_allowed_protocols(self.raw)
        })
    }

    /// Sets the corresponding `TKSmartCard` value.
    pub fn set_allowed_protocols(&self, protocols: SmartCardProtocol) {
        unsafe {
            ffi::smart_card::ctk_smart_card_set_allowed_protocols(self.raw, protocols.bits());
        };
    }

    #[must_use]
    /// Wraps the corresponding `TKSmartCard` operation.
    pub fn current_protocol(&self) -> SmartCardProtocol {
        SmartCardProtocol::from_bits(unsafe {
            ffi::smart_card::ctk_smart_card_current_protocol(self.raw)
        })
    }

    #[must_use]
    /// Wraps the corresponding `TKSmartCard` operation.
    pub fn sensitive(&self) -> bool {
        unsafe { ffi::smart_card::ctk_smart_card_sensitive(self.raw) }
    }

    /// Sets the corresponding `TKSmartCard` value.
    pub fn set_sensitive(&self, sensitive: bool) {
        unsafe { ffi::smart_card::ctk_smart_card_set_sensitive(self.raw, sensitive) };
    }

    #[must_use]
    /// Wraps the corresponding `TKSmartCard` operation.
    pub fn cla(&self) -> u8 {
        unsafe { ffi::smart_card::ctk_smart_card_cla(self.raw) }
    }

    /// Sets the corresponding `TKSmartCard` value.
    pub fn set_cla(&self, cla: u8) {
        unsafe { ffi::smart_card::ctk_smart_card_set_cla(self.raw, cla) };
    }

    #[must_use]
    /// Wraps the corresponding `TKSmartCard` operation.
    pub fn use_extended_length(&self) -> bool {
        unsafe { ffi::smart_card::ctk_smart_card_use_extended_length(self.raw) }
    }

    /// Sets the corresponding `TKSmartCard` value.
    pub fn set_use_extended_length(&self, enabled: bool) {
        unsafe { ffi::smart_card::ctk_smart_card_set_use_extended_length(self.raw, enabled) };
    }

    #[must_use]
    /// Wraps the corresponding `TKSmartCard` operation.
    pub fn use_command_chaining(&self) -> bool {
        unsafe { ffi::smart_card::ctk_smart_card_use_command_chaining(self.raw) }
    }

    /// Sets the corresponding `TKSmartCard` value.
    pub fn set_use_command_chaining(&self, enabled: bool) {
        unsafe { ffi::smart_card::ctk_smart_card_set_use_command_chaining(self.raw, enabled) };
    }

    /// Wraps the corresponding `TKSmartCard` operation.
    pub fn context(&self) -> Result<Option<String>, CryptoTokenKitError> {
        let ptr = unsafe { ffi::smart_card::ctk_smart_card_context_json(self.raw) };
        if ptr.is_null() {
            return Ok(None);
        }
        Ok(Some(crate::error::take_owned_c_string(ptr)))
    }

    /// Sets the corresponding `TKSmartCard` value.
    pub fn set_context(&self, json: Option<&str>) -> Result<(), CryptoTokenKitError> {
        let mut error_ptr = ptr::null_mut();
        let (status, _storage) = if let Some(json) = json {
            let storage = crate::private::to_cstring(json)?;
            let status = unsafe {
                ffi::smart_card::ctk_smart_card_set_context_json(
                    self.raw,
                    storage.as_ptr(),
                    true,
                    &mut error_ptr,
                )
            };
            (status, Some(storage))
        } else {
            let status = unsafe {
                ffi::smart_card::ctk_smart_card_set_context_json(
                    self.raw,
                    ptr::null(),
                    false,
                    &mut error_ptr,
                )
            };
            (status, None)
        };
        status_result(status, error_ptr)
    }

    /// Invokes the corresponding `TKSmartCard` operation.
    pub fn begin_session(&self) -> Result<(), CryptoTokenKitError> {
        let mut error_ptr = ptr::null_mut();
        let status =
            unsafe { ffi::smart_card::ctk_smart_card_begin_session(self.raw, &mut error_ptr) };
        status_result(status, error_ptr)
    }

    /// Wraps the corresponding `TKSmartCard` operation.
    pub fn transmit_request(&self, request: &[u8]) -> Result<Vec<u8>, CryptoTokenKitError> {
        let mut error_ptr = ptr::null_mut();
        let mut reply_ptr = ptr::null_mut();
        let status = unsafe {
            ffi::smart_card::ctk_smart_card_transmit_request_json(
                self.raw,
                request.as_ptr(),
                request.len(),
                &mut reply_ptr,
                &mut error_ptr,
            )
        };
        status_result(status, error_ptr)?;
        if reply_ptr.is_null() {
            return Err(CryptoTokenKitError::FrameworkError(
                "Swift bridge returned a null transmit reply".into(),
            ));
        }
        decode_json(reply_ptr)
    }

    /// Invokes the corresponding `TKSmartCard` operation.
    pub fn end_session(&self) {
        unsafe { ffi::smart_card::ctk_smart_card_end_session(self.raw) };
    }

    /// Wraps the corresponding `TKSmartCard` operation.
    pub fn send_ins(
        &self,
        ins: u8,
        p1: u8,
        p2: u8,
        data: Option<&[u8]>,
        le: Option<usize>,
    ) -> Result<ApduResponse, CryptoTokenKitError> {
        let (data_ptr, data_len) =
            data.map_or((ptr::null(), 0), |bytes| (bytes.as_ptr(), bytes.len()));
        let mut reply_ptr = ptr::null_mut();
        let mut error_ptr = ptr::null_mut();
        let status = unsafe {
            ffi::smart_card::ctk_smart_card_send_ins(
                self.raw,
                ins,
                p1,
                p2,
                data_ptr,
                data_len,
                le.is_some(),
                le.unwrap_or_default(),
                &mut reply_ptr,
                &mut error_ptr,
            )
        };
        status_result(status, error_ptr)?;
        if reply_ptr.is_null() {
            return Err(CryptoTokenKitError::FrameworkError(
                "Swift bridge returned a null APDU reply".into(),
            ));
        }
        decode_json(reply_ptr)
    }

    /// Runs the corresponding `TKSmartCard` workflow around the provided callback.
    pub fn with_session<T>(
        &self,
        callback: impl FnOnce(&Self) -> Result<T, CryptoTokenKitError>,
    ) -> Result<T, CryptoTokenKitError> {
        self.begin_session()?;
        let result = callback(self);
        self.end_session();
        result
    }
}

impl Drop for SmartCard {
    fn drop(&mut self) {
        if !self.raw.is_null() {
            unsafe { ffi::ctk_object_release(self.raw) };
            self.raw = ptr::null_mut();
        }
    }
}
